### PR TITLE
Fix permissions on `keccak.md`

### PR DESCRIPTION
**Pull request by palinatolmach**
Fixes an issue introduced in https://github.com/runtimeverification/kontrol/pull/779 by making `out/kompiled/requires/keccak.md` (and any other filed copied to `requires` in future) writeable for the current user, group, and others (anyone with access to the file).

The files have to writeable so they can be deleted and replaced when `kontrol build --rekompile` is executed. The files not copied from `kontrol/kdist` had write access previously, but the ones we are now copying from `kdist` (such as `keccak.md`) don't, which  causes issues such as this one:

```
INFO 2024-08-23 08:58:19,927 kontrol.kompile - Copying requires path: /nix/store/g67pca4v4lx9zgna190kdwvrj4n0jarm-python3.10-kontrol-0.1.409/lib/python3.10/site-packages/kontrol/kdist/keccak.md ->                           
out/kompiled/requires/keccak.md                                                                                                                                                                                                
An error occurred while building your Kontrol project: [Errno 13] Permission denied: 'out/kompiled/requires/keccak.md'
```

### CODE REVIEW

File: src/kontrol/kompile.py [hunk 1]
```diff
@@ -75,6 +77,12 @@ def foundry_kompile(
         if regen or not req_path.exists():
             _LOGGER.info(f'Copying requires path: {req} -> {req_path}')
             shutil.copy(req, req_path)
+            # If the copied file is not writeable
+            if not os.access(req_path, os.W_OK):
+                # Fetch current permissions
+                current_permissions = os.stat(req_path).st_mode
+                # Grant write permissions
+                os.chmod(req_path, current_permissions | stat.S_IWUSR | stat.S_IWGRP | stat.S_IWOTH)
```
Comment (tothtamas28) [inline src/kontrol/kompile.py:85]: Why is it necessary to have write access on these requires?

Comment (palinatolmach) [reply src/kontrol/kompile.py:85]: We want to be able to delete and replace them when `kontrol build --rekompile` is executed. The files not copied from `kontrol/kdist` had write access previously, but the we are not copying from `kdist` (such as `keccak.md`) don't, and it causes issues such as this one:
```
INFO 2024-08-23 08:58:19,927 kontrol.kompile - Copying requires path: /nix/store/g67pca4v4lx9zgna190kdwvrj4n0jarm-python3.10-kontrol-0.1.409/lib/python3.10/site-packages/kontrol/kdist/keccak.md ->                           
out/kompiled/requires/keccak.md                                                                                                                                                                                                
An error occurred while building your Kontrol project: [Errno 13] Permission denied: 'out/kompiled/requires/keccak.md'
```

Comment (tothtamas28) [reply src/kontrol/kompile.py:85]: > it causes issues such as this one

Please include the issue in the PR description.

Comment (tothtamas28) [inline src/kontrol/kompile.py:85]: Consider using `Path.chmod`.

File: src/kontrol/kompile.py [hunk 2]
```diff
@@ -75,6 +77,12 @@ def foundry_kompile(
         if regen or not req_path.exists():
             _LOGGER.info(f'Copying requires path: {req} -> {req_path}')
             shutil.copy(req, req_path)
+            # If the copied file is not writeable
+            if not os.access(req_path, os.W_OK):
+                # Fetch current permissions
+                current_permissions = os.stat(req_path).st_mode
```
Comment (tothtamas28) [inline src/kontrol/kompile.py:83]: Consider using `Path.stat`.

**Review by tothtamas28 — APPROVED** [review]

File: src/kontrol/kompile.py (hunk 1 above)
Comment (palinatolmach) [reply src/kontrol/kompile.py:85]: Done, thank you!

**Comment by tothtamas28** [comment]
Btw, is it possible the root of the issue is that these files have not been checked in to source control with the right permission?

File: src/kontrol/kompile.py (hunk 1 above)
Comment (palinatolmach) [reply src/kontrol/kompile.py]: Done in https://github.com/runtimeverification/kontrol/pull/785/commits/597820377463129484ee18bd5a52df6855957d09, thanks!

File: src/kontrol/kompile.py (hunk 2 above)
Comment (palinatolmach) [reply src/kontrol/kompile.py]: Done in https://github.com/runtimeverification/kontrol/pull/785/commits/597820377463129484ee18bd5a52df6855957d09, thanks!

File: src/kontrol/kompile.py [hunk 3]
```diff
@@ -75,6 +76,12 @@ def foundry_kompile(
         if regen or not req_path.exists():
             _LOGGER.info(f'Copying requires path: {req} -> {req_path}')
             shutil.copy(req, req_path)
+            # If the copied file is not writeable
+            if not os.access(req_path, os.W_OK):
+                # Fetch current permissions
+                current_permissions = Path.stat(req_path).st_mode
+                # Grant write permissions
+                Path.chmod(req_path, current_permissions | Path.stat.S_IWUSR | Path.stat.S_IWGRP | Path.stat.S_IWOTH)
```
Comment (tothtamas28) [inline src/kontrol/kompile.py:84]:
```suggestion
                req_path.chmod(current_permissions | Path.stat.S_IWUSR | Path.stat.S_IWGRP | Path.stat.S_IWOTH)
```

File: src/kontrol/kompile.py [hunk 4]
```diff
@@ -75,6 +76,12 @@ def foundry_kompile(
         if regen or not req_path.exists():
             _LOGGER.info(f'Copying requires path: {req} -> {req_path}')
             shutil.copy(req, req_path)
+            # If the copied file is not writeable
+            if not os.access(req_path, os.W_OK):
+                # Fetch current permissions
+                current_permissions = Path.stat(req_path).st_mode
```
Comment (tothtamas28) [inline src/kontrol/kompile.py:82]:
```suggestion
                current_permissions = req_path.stat().st_mode
```

**Comment by palinatolmach** [comment]
@tothtamas28 it would probably help, but these files are part of the kompilation artifacts usually placed in the `/out` folder in Foundry projects, and it's usually not tracked by source control at all (here's [one example](https://github.com/morpho-org/morpho-blue/blob/8e35224dfc69eebf68ad67408bb1154173aeed16/.gitignore#L3) from a project we're looking at).

File: src/kontrol/kompile.py (hunk 3 above)
Comment (palinatolmach) [reply src/kontrol/kompile.py:84]: Oh, sorry, I misunderstood how it should be used. Thank you!

**Comment by tothtamas28** [comment]
> these files are part of the kompilation artifacts usually placed in the `/out` folder in Foundry projects

What I'm wondering, if the template of these files (i.e. in the `KSRC_DIR`) have the correct permissions, and [`shutil,copy`](https://docs.python.org/3/library/shutil.html#shutil.copy) preserves file permissions, why do the files placed in `out` have the wrong permissions.

**Comment by palinatolmach** [comment]
> What I'm wondering, if the template of these files (i.e. in the KSRC_DIR) have the correct permissions, and [shutil,copy](https://docs.python.org/3/library/shutil.html#shutil.copy) preserves file permissions, why do the files placed in out have the wrong permissions.

Ah, got it, thanks, it looks like they don't.

If `kontrol build` is run from a local checkout (`poetry run kontrol build`), then these files are copied with write permissions as they are in `src/kontrol/kdist`:
```
❯ ls -l src/kontrol/kdist
total 304
-rw-r--r--  1 palina  staff      0 Apr  9 20:53 __init__.py
drwxr-xr-x  5 palina  staff    160 Apr 11 16:36 __pycache__
-rw-r--r--  1 palina  staff  19147 Aug  9 15:15 assert.md
...
-rw-r--r--@ 1 palina  staff   3218 Aug 22 22:15 keccak.md
```

If we run `kontrol build` using a `kup` installation, then they're copied from something like `/nix/store/gyx8hr1kq30chx12bdziqv5yyz582ql3-python3.10-kontrol-0.1.409/lib/python3.10/site-packages/kontrol/kdist/`, and the `kdist` files only have read permissions there:
```
> ls -l /nix/store/gyx8hr1kq30chx12bdziqv5yyz582ql3-python3.10-kontrol-0.1.409/lib/python3.10/site-packages/kontrol/kdist/
-r--r--r--@ 1 root  wheel      0 Jan  1  1970 __init__.py
dr-xr-xr-x@ 8 root  wheel    256 Jan  1  1970 __pycache__
-r--r--r--@ 1 root  wheel  19147 Jan  1  1970 assert.md
...
-r--r--r--@ 1 root  wheel   3218 Jan  1  1970 keccak.md
```

In short, it seems that the `kdist` files distributed with `kup`-installed Kontrol are read-only, and I thought it's intentional.

**Comment by tothtamas28** [comment]
> In short, it seems that the `kdist` files distributed with `kup`-installed Kontrol are read-only, and I thought it's intentional.

Makes sense, thank you for debugging this. The Nix store in general is read only, so your fix is indeed the right thing to do.

**Comment by palinatolmach** [comment]
@tothtamas28 thanks for your help!

**Review by anvacaru — APPROVED** [review]
